ktor-features/ktor-serialization/jvm/src/io/ktor/serialization/SerializerLookup.kt

```diff
@@ -36,33 +38,49 @@ private fun arraySerializer(type: KType): KSerializer<*> {
 }
 
 @OptIn(InternalSerializationApi::class)
-internal fun serializerForSending(value: Any, module: SerializersModule): KSerializer<*> = when (value) {
-    is JsonElement -> JsonElement.serializer()
-    is List<*> -> ListSerializer(value.elementSerializer(module))
-    is Set<*> -> SetSerializer(value.elementSerializer(module))
-    is Map<*, *> -> MapSerializer(value.keys.elementSerializer(module), value.values.elementSerializer(module))
-    is Map.Entry<*, *> -> MapEntrySerializer(
-        serializerForSending(value.key ?: error("Map.Entry(null, ...) is not supported"), module),
-        serializerForSending(value.value ?: error("Map.Entry(..., null) is not supported)"), module)
-    )
-    is Array<*> -> {
-        val componentType = value.javaClass.componentType.kotlin.starProjectedType
-        val componentClass =
-            componentType.classifier as? KClass<*> ?: error("Unsupported component type $componentType")
+internal fun serializerForSending(
+    context: PipelineContext<Any, ApplicationCall>,
+    value: Any,
+    module: SerializersModule
+): KSerializer<*> {
+    val responseType = context.call.attributes.getOrNull(ResponseTypeAttributeKey)
+    if (responseType != null) try {
+        return module.serializer(responseType)
+    } catch (_: SerializationException) {
+        // if serializer for type was not found, fallback to manual search
```

Comment:
When I don't like here is that all the default content types will always go through the exception. This is good and bad at the same time.
- Pros: one could configure serialization for types like `HttpStatusCode`
- Cons: looking for a serializer, creating an exception instance are slow

Reply:
Unfortunately, there is no better way to check if a serializer is available. I created the issue in `kotlinx.serializer` project https://github.com/Kotlin/kotlinx.serialization/issues/1163

Reply:
I updated my code to create `KType` only when the user responds with a custom type